shared/src/ledger/eth_bridge/storage/eth_msgs.rs

```diff
@@ -0,0 +1,199 @@
+#[allow(missing_docs)]
+pub const SEEN_BY_KEY_SEGMENT: &str = "seen_by";
+#[allow(missing_docs)]
+pub const VOTING_POWER_KEY_SEGMENT: &str = "voting_power";
+
+/// Generator for the keys under which details of an [`EthereumEvent`] is stored
+pub struct Keys {
+    /// The prefix under which the details are an [`EthereumEvent`] is stored
+    pub prefix: Key,
+}
+
+impl Keys {
+    /// Get the `body` key- there should be an [`EthereumEvent`] stored here.
+    pub fn body(&self) -> Key {
+        self.prefix
+            .push(&BODY_KEY_SEGMENT.to_owned())
+            .expect("should always be able to construct this key")
+    }
+
+    /// Get the `seen` key - there should be a [`bool`] stored here.
+    pub fn seen(&self) -> Key {
+        self.prefix
+            .push(&SEEN_KEY_SEGMENT.to_owned())
+            .expect("should always be able to construct this key")
+    }
+
+    /// Get the `seen_by` key - there should be a [`Vec<Address>`] stored here.
+    pub fn seen_by(&self) -> Key {
+        self.prefix
+            .push(&SEEN_BY_KEY_SEGMENT.to_owned())
+            .expect("should always be able to construct this key")
+    }
+
+    /// Get the `voting_power` key - there should be a `(u64, u64)` stored
+    /// here.
+    pub fn voting_power(&self) -> Key {
+        self.prefix
+            .push(&VOTING_POWER_KEY_SEGMENT.to_owned())
+            .expect("should always be able to construct this key")
+    }
+}
+
+impl IntoIterator for &Keys {
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+    type Item = Key;
+
+    fn into_iter(self) -> Self::IntoIter {
+        vec![
+            self.body(),
+            self.seen(),
+            self.seen_by(),
+            self.voting_power(),
+        ]
+        .into_iter()
+    }
+}
+
+impl From<&EthereumEvent> for Keys {
+    fn from(event: &EthereumEvent) -> Self {
+        let hash = event
+            .hash()
+            .expect("should always be able to hash Ethereum events");
+        (&hash).into()
+    }
+}
+
+impl From<&Hash> for Keys {
+    fn from(hash: &Hash) -> Self {
+        let hex = format!("{}", hash);
+        let prefix = prefix()
+            .push(&hex)
+            .expect("should always be able to construct this key");
+        Self { prefix }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::ledger::eth_bridge::ADDRESS;
+    use crate::types::storage::DbKeySeg;
+
+    fn arbitrary_event_with_hash() -> (EthereumEvent, String) {
+        (
+            EthereumEvent::TransfersToNamada {
+                nonce: 1.into(),
+                transfers: vec![],
+            },
+            "06799912C0FD8785EE29E13DFB84FE2778AF6D9CA026BD5B054F86CE9FE8C017"
+                .to_owned(),
+        )
+    }
+
+    #[test]
+    fn test_prefix() {
+        assert_matches!(&prefix().segments[..], [
+                DbKeySeg::AddressSeg(ADDRESS),
+                DbKeySeg::StringSeg(s),
+            ] if s == PREFIX_KEY_SEGMENT)
```

Comment:
I think you can pattern match on `DbKeySeg::StringSeg(PREFIX_KEY_SEGMENT)`directly, kind of like you did for `DbKeySeg::AddressSeg(ADDRESS)`

Reply:
I'm not sure how, this doesn't work?

```rust
        assert_matches!(
            &prefix().segments[..],
            [
                DbKeySeg::AddressSeg(ADDRESS),
                DbKeySeg::StringSeg(PREFIX_KEY_SEGMENT.to_owned()),
            ]
        );
```

Reply:
oh it's an owned string. nevermind then!